floors/fetcher.go

```diff
@@ -0,0 +1,323 @@
+	if config.Enabled && config.Fetcher.Enabled && config.Fetcher.Timeout > 0 {
+		fetchInfo := FetchInfo{AccountFloorFetch: config.Fetcher, FetchTime: f.time.Now().Unix(), RefetchRequest: false}
+		f.configReceiver <- fetchInfo
+	}
+
+	return nil, openrtb_ext.FetchInprogress
+}
+
+func (f *PriceFloorFetcher) worker(config config.AccountFloorFetch) {
+	floorData, fetchedMaxAge := f.fetchAndValidate(config)
+	if floorData != nil {
+		// Update cache with new floor rules
+		glog.Infof("Updating Value in cache for URL %s", config.URL)
+		cacheExpiry := config.MaxAge
+		if fetchedMaxAge != 0 && fetchedMaxAge > config.Period && fetchedMaxAge < math.MaxInt32 {
+			cacheExpiry = fetchedMaxAge
+		}
+		floorData, err := json.Marshal(floorData)
+		if err != nil {
+			glog.Errorf("Error while marshaling fetched floor data for url %s", config.URL)
+		} else {
+			f.SetWithExpiry(config.URL, floorData, cacheExpiry)
+		}
+	}
+
+	// Send to refetch channel
+	f.configReceiver <- FetchInfo{AccountFloorFetch: config, FetchTime: f.time.Now().Add(time.Duration(config.Period) * time.Second).Unix(), RefetchRequest: true}
+
+}
+
+// Stop terminates fetcher service can closes all the channels and destroys worker pool
+func (f *PriceFloorFetcher) Stop() {
+	if f == nil {
+		return
+	}
+
+	close(f.done)
+	f.pool.Stop()
+	close(f.configReceiver)
+}
+
+func (f *PriceFloorFetcher) submit(fetchInfo *FetchInfo) {
+	status := f.pool.TrySubmit(func() {
+		f.worker(fetchInfo.AccountFloorFetch)
+	})
+	if !status {
+		heap.Push(&f.fetchQueue, fetchInfo)
+	}
+}
+
+func (f *PriceFloorFetcher) Fetcher() {
+	//Create Ticker of 5 minutes
+	ticker := time.NewTicker(time.Duration(refetchCheckInterval) * time.Second)
+
+	for {
+		select {
+		case fetchInfo := <-f.configReceiver:
+			if fetchInfo.RefetchRequest {
+				heap.Push(&f.fetchQueue, &fetchInfo)
+			} else {
+				if _, ok := f.fetchInProgress[fetchInfo.URL]; !ok {
+					f.fetchInProgress[fetchInfo.URL] = true
+					f.submit(&fetchInfo)
+				}
+			}
+		case <-ticker.C:
+			currentTime := f.time.Now().Unix()
+			for top := f.fetchQueue.Top(); top != nil && top.FetchTime <= currentTime; top = f.fetchQueue.Top() {
+				nextFetch := heap.Pop(&f.fetchQueue)
+				f.submit(nextFetch.(*FetchInfo))
+			}
+		case <-f.done:
+			glog.Info("Price Floor fetcher terminated")
+			return
+		}
+	}
+}
+
+func (f *PriceFloorFetcher) fetchAndValidate(config config.AccountFloorFetch) (*openrtb_ext.PriceFloorRules, int) {
+	floorResp, maxAge, err := f.fetchFloorRulesFromURL(config)
+	if err != nil {
+		glog.Errorf("Error while fetching floor data from URL: %s, reason : %s", config.URL, err.Error())
+		return nil, 0
+	}
+
+	if len(floorResp) > (config.MaxFileSizeKB * 1024) {
+		glog.Errorf("Recieved invalid floor data from URL: %s, reason : floor file size is greater than MaxFileSize", config.URL)
+		return nil, 0
+	}
+
+	var priceFloors openrtb_ext.PriceFloorRules
+	if err = json.Unmarshal(floorResp, &priceFloors.Data); err != nil {
```

Comment:
This was hard to spot. Only noticed until I tested locally. We are trying to unmarshal into `&priceFloors` not `&priceFloors.Data`

Reply:
Actually, only data object will be hosted not entire floor json
It is mentioned in the PRD

Reply:
```
When data is retrieved from a dynamic fetch, it must be assumed to be entirely under the 'data' element of Schema 2.
As described in the signaling section below, fetch data will be merged into the floors request config under 'data' before validating. 
The idea is that the request/storedrequest will define some pieces of the overall floor config
(e.g.floorMin, enabled) and the floor provider supplies the actual floor rules under the 'data' section.
```
From PRD